src/platform_impl/web/window.rs

```diff
@@ -60,42 +60,47 @@ impl Window {
         let destroy_fn = Box::new(move || runner.notify_destroy_window(RootWI(id)));
 
         let has_focus = canvas.borrow().has_focus.clone();
-        let window = Window {
+        let inner = Inner {
             id,
+            window: window.clone(),
+            document: document.clone(),
+            canvas,
+            previous_pointer: RefCell::new("auto"),
+            register_redraw_request,
+            destroy_fn: Some(destroy_fn),
             has_focus,
-            inner: Dispatcher::new(Inner {
-                window: window.clone(),
-                document: document.clone(),
-                canvas,
-                previous_pointer: RefCell::new("auto"),
-                register_redraw_request,
-                destroy_fn: Some(destroy_fn),
-            })
-            .unwrap(),
         };
 
-        window.set_title(&attr.title);
-        window.set_maximized(attr.maximized);
-        window.set_visible(attr.visible);
-        window.set_window_icon(attr.window_icon);
+        inner.set_title(&attr.title);
+        inner.set_maximized(attr.maximized);
+        inner.set_visible(attr.visible);
+        inner.set_window_icon(attr.window_icon);
+
+        Ok(Window {
+            inner: Dispatcher::new(inner).unwrap(),
+        })
+    }
 
-        Ok(window)
+    pub(crate) fn maybe_queue_on_main(&self, f: impl FnOnce(&Inner) + Send + 'static) {
+        self.inner.dispatch(f)
+    }
+
+    // TODO: Find a way to make this non-static
+    pub(crate) fn maybe_wait_on_main<R: Send + 'static>(
+        &self,
+        f: impl FnOnce(&Inner) -> R + Send + 'static,
+    ) -> R {
+        self.inner.queue(f)
```

Comment:
I am unsure if these two are the best abstractions?

Although slightly less efficient, maybe we should just always use `maybe_wait_on_main`, since that is the only real way to ensure that an action completed, and to propagate panics? I fear that the "fire and forget" style of programming that `maybe_queue_on_main` is will only lead to problems down the road.

Reply:
I don't mind not propagating the panics into the calling thread.

At least on Web we only panic if our assumptions are wrong, which should never happen unless Winit or the browser has a bug. Generally speaking panicking on Web right now is problematic, it should be considered unrecoverable until the exception handling proposal is implemented (which was recently done, but only for `no_std`, which we don't support for Web).

I do agree though that this isn't ideal, but I would like to avoid blocking the thread when we can, so I would argue it's worth it. Though we could at least add some documentation noting this behavior somewhere. I suggest we add it to the `Window` type docs, where we should have noted that some of these function can block in a multi-threaded environment anyway.

Reply:
I've added some (kinda vague) docs about this to `Window`, so that the user is aware that things might block, but not enough that we can't change internally which things block.

Please proof-read it!